### DISCUSSION TITLE
Functions v4 Breaking Change Process

### DISCUSSION THREAD

**Discussion by brettsam** [Azure/Azure-Functions#1909] [Design discussions]
With Functions v4 (running on .NET 6.0), we want to take a very deliberate approach to breaking changes. We're requiring that all potential changes are proposed and approved. We've also created a template to guide these proposals and ensure that most scenarios are considered before presenting.

It is **not** a goal that a customer can change their site extension version from ~3 to ~4 and be guaranteed to run without errors. However, we want to have as smooth as an upgrade path as possible. This includes:
  - Breaking change changelog file checked into this meta repo.
  - Release notes and documentation.
  - Migration guides.
  - Backwards compatibility [feature flags](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#azurewebjobsfeatureflags) (where possible).
     - For example, in v3 we no longer allow synchronous IO by default. However by setting the `AllowSynchronousIO` feature flag, it will be enabled again.

Breaking changes will not be accepted after some future date (TBD), to allow for stabilization before GA.

### Breaking Change Approval Process
The approval process looks like:
  1. Complete the [Functions v4 Breaking Change Issue Template](https://github.com/Azure/Azure-Functions/issues/new?assignees=&labels=&template=breaking_change_proposal.md) as thoroughly as possible.
      - Mention @brettsam and @fabiocav in the issue so we are aware of it.
      - Click **Lock conversation** on the issue immediately after creating it. If you do not have permissions, one of the team members can do this.
        All discussions for a proposal will happen in github disscussions and only updates/approvals/etc will be made in the original issue. This allows for customers to follow the issue without getting notified of every reply to the discussion.
  2. Create a [New discussion](https://github.com/Azure/Azure-Functions/discussions/new) with the same title as the issue. The category for the discussion should be `Design discussions`.
  3. Link the discussion to the issue, and vice versa.
  4. The Functions team will use the discussion to ask questions, provide suggestions, etc.
  5. Once approved, the work will be assigned.
  6. Once complete, include with your PR an update the Breaking Changes changelog file (details TBD).

**Comment by Hazhzeng** [comment]
I cannot select discussion category which blocks me from creating a new discussion.
![fail](https://user-images.githubusercontent.com/48038149/125524776-39bb263e-a495-4782-9421-4991bc5c414c.gif)

**Comment by fabiocav** [comment]
This is a bug in GitHub permissions.

The easiest way is for me to create the discussion if you’d like.

From: Hanzhang Zeng ***@***.***>
Date: Tuesday, July 13, 2021 at 2:04 PM
To: Azure/Azure-Functions ***@***.***>
Cc: Fabio Cavalcante ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)

I cannot select discussion category which blocks me from creating a new discussion.
[fail]<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F48038149%2F125524776-39bb263e-a495-4782-9421-4991bc5c414c.gif&data=04%7C01%7Cfacaval%40microsoft.com%7C5524894f4b884009e64108d94641d806%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637618070884701745%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=r1RL9dlK4YKD4p4rsBeDnTqVHCa6j%2F%2B%2BNqxvDrXjZy4%3D&reserved=0>

**Comment by fabiocav** [comment]
Should the title match what you have below?

From: Fabio Cavalcante ***@***.***>
Date: Tuesday, July 13, 2021 at 2:57 PM
To: Azure/Azure-Functions ***@***.***>, Azure/Azure-Functions ***@***.***>
Cc: Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)
This is a bug in GitHub permissions.

The easiest way is for me to create the discussion if you’d like.

From: Hanzhang Zeng ***@***.***>
Date: Tuesday, July 13, 2021 at 2:04 PM
To: Azure/Azure-Functions ***@***.***>
Cc: Fabio Cavalcante ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)

I cannot select discussion category which blocks me from creating a new discussion.
[fail]<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F48038149%2F125524776-39bb263e-a495-4782-9421-4991bc5c414c.gif&data=04%7C01%7Cfacaval%40microsoft.com%7C5524894f4b884009e64108d94641d806%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637618070884701745%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=r1RL9dlK4YKD4p4rsBeDnTqVHCa6j%2F%2B%2BNqxvDrXjZy4%3D&reserved=0>

**Comment by fabiocav** [comment]
Actually, looks like @***@***.***> already set this up here: Remove thread count limitation on workers · Discussion #1967 · Azure/Azure-Functions (github.com)<https://github.com/Azure/Azure-Functions/discussions/1967>


From: Fabio Cavalcante ***@***.***>
Date: Tuesday, July 13, 2021 at 2:58 PM
To: Azure/Azure-Functions ***@***.***>, Azure/Azure-Functions ***@***.***>
Cc: Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)
Should the title match what you have below?

From: Fabio Cavalcante ***@***.***>
Date: Tuesday, July 13, 2021 at 2:57 PM
To: Azure/Azure-Functions ***@***.***>, Azure/Azure-Functions ***@***.***>
Cc: Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)
This is a bug in GitHub permissions.

The easiest way is for me to create the discussion if you’d like.

From: Hanzhang Zeng ***@***.***>
Date: Tuesday, July 13, 2021 at 2:04 PM
To: Azure/Azure-Functions ***@***.***>
Cc: Fabio Cavalcante ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/Azure-Functions] Functions v4 Breaking Change Process (#1909)

I cannot select discussion category which blocks me from creating a new discussion.
[fail]<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F48038149%2F125524776-39bb263e-a495-4782-9421-4991bc5c414c.gif&data=04%7C01%7Cfacaval%40microsoft.com%7C5524894f4b884009e64108d94641d806%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637618070884701745%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=r1RL9dlK4YKD4p4rsBeDnTqVHCa6j%2F%2B%2BNqxvDrXjZy4%3D&reserved=0>

**Comment by Hazhzeng** [comment]
Yeah, I think Brett already created one for me. Thanks Fabio. https://github.com/Azure/Azure-Functions/discussions/1967
